File: dynamic_programming/#152_maximum_product_subarray/maxProduct.go
package main

import (
	"fmt"
)

func main() {
	maxProduct([]int{2, 3, -2, 4})     // 6
	maxProduct([]int{2, 3, -2, 4, -2}) // 96
	maxProduct([]int{-2, 0, -1})       // 0
}

func maxProduct(nums []int) int {
	curr_max, curr_min, final_max := nums[0], nums[0], nums[0]

	for _, val := range nums[1:] {
		next_max := curr_max * val
		next_mix := curr_min * val
		curr_max = maxOf(val, next_max, next_mix)
		curr_min = minOf(val, next_max, next_mix)
		final_max = maxOf(final_max, curr_max)
	}

	fmt.Println(final_max)
	return final_max
}

func minOf(vars ...int) int {
	min := vars[0]

	for _, val := range vars {
		if val < min {
			min = val
		}
	}

	return min
}

func maxOf(vars ...int) int {
	max := vars[0]

	for _, val := range vars {
		if val > max {
			max = val
		}
	}

	return max
}

File: dynamic_programming/#152_maximum_product_subarray/maxProduct.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function (nums) { // TC: O(n), SC: O(1)
  let curr_max = nums[0], curr_min = nums[0], final_max = nums[0]

  for (let i = 1; i < nums.length; i++) {
    const num = nums[i]
    next_max = curr_max * num // 下一個的可能最大
    next_min = curr_min * num // 下一個的可能最小
    curr_max = Math.max(next_max, next_min, num)
    curr_min = Math.min(next_max, next_min, num)
    final_max = Math.max(final_max, curr_max)
  }

  console.log(final_max);
  return final_max
};

maxProduct([2, 3, -2, 4]) // 6
maxProduct([2, 3, -2, 4, -2]) // 96
maxProduct([-2, 0, -1]) // 0